core/primitives-core/src/config.rs

```diff
@@ -242,10 +242,10 @@ pub struct ExtCostsConfig {
     /// Base cost for calling a host function.
     pub base: Gas,
 
-    /// Base cost of loading and compiling contract
-    pub contract_compile_base: Gas,
-    /// Cost of the execution to load and compile contract
-    pub contract_compile_bytes: Gas,
```

Comment:
@matklad @bowenwang1996 I am not sure how to proceed now, but there is quite unfortunate API breaking change.

> I created this as a separate PR to make sure we can consider carefully that this does not have any unwanted side-effects at some distant place.

This actually broke JSON RPC API (cc @miraclx @khorolets) compatibility since previously there was `contract_compile_base` field in `EXPERIMENTAL_protocol_config` API, and now there is no such field (new fields do not cause any problems on JSON deserialization, but missing fields do).

Apps that use near-jsonrpc-client-rs (which uses `near-primitives-core` published on creates.io) cannot use the `EXPERIMENTAL_protocol_config` API on testnet at the moment (mainnet doesn't have this change yet):

```
RpcError: RpcError { error_struct: Some(RequestValidationError(ParseError { error_message: "Failed to parse: Error(\"missing field `contract_compile_base`\", line: 0, column: 0)" })), code: -32700, message: "Parse error", data: Some(String("Failed to parse: Error(\"missing field `contract_compile_base`\", line: 0, column: 0)")) }
```

I also believe it broke the `tx` JSON RPC API where `gas_profile` is the list of costs where `CONTRACT_COMPILE_BASE` identifier was used:

<img width="1053" alt="image" src="https://user-images.githubusercontent.com/304265/169095635-4e06b123-86d6-408c-aedc-cdbb670e6fa6.png">

Reply:
>I also believe it broke the tx JSON RPC API where gas_profile is the list of costs where CONTRACT_COMPILE_BASE identifier was used:

For gas profile, the JSON type is deliberately `Vec<(String, String, Gas)>`, to make sure we don't hit such kinds of problems. 

https://github.com/near/nearcore/blob/8d2a9fe72f216aab0dfde2b3dc3776de803d76c9/core/primitives/src/views.rs#L1107-L1122

For the config, yeah, that's problematic... 

Am I correct that, as this is a somewhat old change, the code is sufficiently deployed that we need to fix that on the client-side, rather than on the server side? That is, on server-side solution would be to just slap `serde(rename)` on those fields, but would that actually solve anything? 

If the answer to above is "yes", then what we could do is slapping `serde(default)` instead, so that new version of JSON RPC client can read both old and new versions of the config? 

What action item do we have to fix the immediate problem?

---

I would also like to see some long-term plan here. It's indeed quite suboptimal that a  rename in the guts of nearcore can break API clients without us noticing that for more than a month! I think this is a result of three somewhat related, but different issues: 

* we don't have a nearcore-independent crate which specifies our JSON API. 
* we don't have tests which check JSON API stability
* for this specific API endpoint, the de-facto JSON format we use leaks internal impl details and is not future-proofed. There's no RuntimeConfigView. Curiously, I believe @jakmeier's work on parameters might help here?

Reply:
I feel we can still issue a hot-fix for the breaking change as 1.26.1, can't we? @bowenwang1996 @mm-near 

> For gas profile, the JSON type is deliberately `Vec<(String, String, Gas)>`, to make sure we don't hit such kinds of problems.

@matklad It only solves the serialization/deserialization, but technically if someone relied on that specific metric, they will get a non-documented breaking change.

> I would also like to see some long-term plan here. It's indeed quite suboptimal that a rename in the guts of nearcore can break API clients without us noticing that for more than a month! I think this is a result of three somewhat related, but different issues

We should finally extract all the serialization/deserialization from allover the places and have a single location for JSON RPC #5516, and similarly we will have it addressed for Indexer Framework.

Reply:
@frol my two cents:
* It is unfortunate that this has happened and we should have tests to prevent this from happening again in the future
* The endpoint is `EXPERIMENTAL_*` and I believe it means that there is no guarantee on backward compatibility.
* Given that nearcore has already been released and many node operators have updated, it seems to me that it is easier to fix on the client side.

Reply:
> Given that nearcore has already been released and many node operators have updated, it seems to me that it is easier to fix on the client side.

@bowenwang1996 first of all, it is never "easier to fix on the client side" because there are unknown number of clients while there is known number of nodes. Second, the fix will anyway land on nearcore repo since "client side" (at least near-jsonrpc-client-rs) is based on nearcore primitive crates, so we will need Node team to take care of it; can Node team take #5516 any time soon?

Reply:
@frol https://github.com/near/nearcore/pull/6844

Reply:
@frol yeah that makes sense. 
> so we will need Node team to take care of it; can Node team take https://github.com/near/nearcore/issues/5516 any time soon?

Unfortunately not this quarter